src/control/cmd/daos_agent/procmon.go

```diff
@@ -0,0 +1,210 @@
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+// The Government's rights to use, modify, reproduce, release, perform, display,
+// or disclose this software are subject to the terms of the Apache License as
+// provided in Contract No. 8F-30005.
+// Any reproduction of computer software, computer software documentation, or
+// portions thereof marked with this legend must also reproduce the markings.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"syscall"
+	"time"
+
+	"github.com/daos-stack/daos/src/control/drpc"
+	"github.com/daos-stack/daos/src/control/logging"
+)
+
+type procMonRequest struct {
+	// Pid of the process that the request is for
+	pid int32
+	// Whether the message is an attach or disconnect. Uses drpc method identifiers.
+	action drpc.MgmtMethod
+}
+
+type procMonResponse struct {
+	// Pid the response is coming from
+	pid int32
+	// Error indicating why the process "died"
+	err error
+}
+
+type procInfo struct {
+	log       logging.Logger
+	pid       int32
+	ctx       context.Context
+	cancelCtx func()
+	response  chan *procMonResponse
+}
+
+// NewProcInfo creates a new procInfo struct for use by procMon. The context passed
+// in is the top level context which a child context and cancellation routine is generated.
+// The response channel is the passed in by procMon to provide a unified interface
+// in procMon to handle responses.
+func NewProcInfo(ctx context.Context, Log logging.Logger, Pid int32, Response chan *procMonResponse) *procInfo {
+	child, cancel := context.WithCancel(ctx)
+	return &procInfo{
+		log:       Log,
+		pid:       Pid,
+		ctx:       child,
+		cancelCtx: cancel,
+		response:  Response,
+	}
+}
+
+func getProcPidInode(pid int32) (uint64, error) {
+	pidPath := fmt.Sprintf("/proc/%d", pid)
+	info, err := os.Stat(pidPath)
+	if err != nil {
+		return 0, err
+	}
+
+	// Make sure info's underlying interface is correct.
+	stat, ok := info.Sys().(*syscall.Stat_t)
+	if !ok {
+		return 0, fmt.Errorf("Underlying type of FileInfo.Sys() is not Stat_t")
+	}
+	return stat.Ino, nil
+}
+
+func (p *procInfo) sendResponse(pid int32, err error) {
+	response := &procMonResponse{
+		pid: pid,
+		err: err,
+	}
+	p.response <- response
+}
+
+const MonWaitTime = 3 * time.Second
+
+// monitorProcess is the top level interface for monitoring individual processes.
+// the interface is used by procMon to kick off monitoring individual processes
+// under their own child context to allow for terminating individual monitoring routines.
```

Comment:
Should probably combine the top two lines to something like "monitorProcess is used by procMon to kick off monitoring individual processes"

Reply:
Fixed the comment.